File: sdks/typescript/src/wire/tuplespace-proto-wire.ts
// SPDX-License-Identifier: LGPL-2.1-or-later
// Copyright (C) 2025 PlexSpaces Contributors
//
// Protobuf wire for plexspaces.tuplespace.v1 WriteRequest / ReadRequest / ReadResponse
// (subset for WASM host imports). Byte-compatible with:
// - sdks/go/plexspaces/tuplespace_proto_wire.go
// - crates/wasm-runtime simple_component_host (prost)

import {
  appendFixed64LE,
  appendLengthDelimited,
  appendVarint,
  concatBytes,
  readLengthDelimited,
  readVarint,
  skipField,
} from './proto-wire-common.js';

const MIN_INT64 = -9223372036854775808n;
const MAX_INT64 = 9223372036854775807n;

/** Protobuf wire bytes backed by ArrayBuffer (matches host/WASM expectations). */
type WireU8 = Uint8Array<ArrayBuffer>;

function encodeTupleField(v: unknown, allowWildcardStar: boolean): WireU8 {
  if (v === null || v === undefined) {
    return appendVarint(new Uint8Array([0x38]), 1);
  }
  if (typeof v === 'string') {
    if (allowWildcardStar && v === '*') {
      return appendVarint(new Uint8Array([0x38]), 1);
    }
    const enc = new TextEncoder();
    const bytes = new Uint8Array(enc.encode(v));
    let inner = new Uint8Array([0x1a]);
    inner = appendVarint(inner, bytes.length);
    inner = concatBytes(inner, bytes);
    return inner;
  }
  if (typeof v === 'boolean') {
    const inner = new Uint8Array([0x20]);
    return appendVarint(inner, v ? 1 : 0);
  }
  if (typeof v === 'number' && Number.isFinite(v)) {
    const t = Math.trunc(v);
    if (t === v && t >= Number(MIN_INT64) && t <= Number(MAX_INT64)) {
      let inner = new Uint8Array([0x08]);
      inner = appendVarintSigned(inner, t);
      return inner;
    }
    let inner = new Uint8Array([0x11]);
    const tmp = new Uint8Array(8);
    new DataView(tmp.buffer).setFloat64(0, v, true);
    inner = concatBytes(inner, tmp);
    return inner;
  }
  throw new Error(`unsupported tuple field type ${typeof v}`);
}

function appendVarintSigned(buf: Uint8Array, xIn: number): WireU8 {
  let x = BigInt(xIn);
  if (x < 0n) x = BigInt.asUintN(64, x);
  const parts: number[] = [];
  let n = x;
  while (n >= 0x80n) {
    parts.push(Number(n & 0xffn) | 0x80);
    n >>= 7n;
  }
  parts.push(Number(n));
  return concatBytes(buf, new Uint8Array(parts));
}

function encodeTupleFields(tuple: unknown[], allowWildcardStar: boolean): WireU8 {
  let out: WireU8 = new Uint8Array(0);
  for (const el of tuple) {
    const tf = encodeTupleField(el, allowWildcardStar);
    out = appendLengthDelimited(out, 2, tf);
  }
  return out;
}

/** Build WriteRequest protobuf bytes (field 1: tuple with repeated TupleField). */
export function encodeWriteRequest(tuple: unknown[]): Uint8Array {
  const tupleBody = encodeTupleFields(tuple, false);
  return appendLengthDelimited(new Uint8Array(0), 1, tupleBody);
}

/** Build ReadRequest protobuf bytes. */
export function encodeReadRequest(pattern: unknown[], take: boolean, maxResults: number): Uint8Array {
  const templateBody = encodeTupleFields(pattern, true);
  let out = appendLengthDelimited(new Uint8Array(0), 1, templateBody);
  if (take) {
    out = concatBytes(out, new Uint8Array([0x20, 0x01]));
  }
  out = concatBytes(out, new Uint8Array([0x28]));
  out = appendVarint(out, maxResults >>> 0);
  return out;
}

function parseTupleFieldMsg(msg: Uint8Array): unknown {
  let pos = 0;
  let last: unknown = undefined;
  while (pos < msg.length) {
    const { value: tag, n: tn } = readVarint(msg, pos);
    pos += tn;
    const fn = Number(tag >> 3n);
    const wt = Number(tag & 7n);
    if (wt === 0) {
      const { value: v, n: m } = readVarint(msg, pos);
      pos += m;
      if (fn === 1) last = Number(v);
      else if (fn === 4) last = v !== 0n;
      else if (fn === 6 || fn === 7) last = null;
    } else if (wt === 1) {
      if (pos + 8 > msg.length) throw new Error('double underflow');
      const view = new DataView(msg.buffer, msg.byteOffset + pos, 8);
      const d = view.getFloat64(0, true);
      pos += 8;
      if (fn === 2) last = d;
    } else if (wt === 2) {
      const { slice: chunk, nextPos } = readLengthDelimited(msg, pos);
      pos = nextPos;
      if (fn === 3 || fn === 5) {
        last = new TextDecoder('utf-8', { fatal: false }).decode(chunk);
      }
    } else {
      pos = skipField(msg, pos, wt);
    }
  }
  return last;
}

function parseTupleMsg(msg: Uint8Array): unknown[] {
  const fields: unknown[] = [];
  let pos = 0;
  while (pos < msg.length) {
    const { value: tag, n: tn } = readVarint(msg, pos);
    pos += tn;
    const fn = Number(tag >> 3n);
    const wt = Number(tag & 7n);
    if (fn === 2 && wt === 2) {
      const { slice: sub, nextPos } = readLengthDelimited(msg, pos);
      pos = nextPos;
      fields.push(parseTupleFieldMsg(sub));
    } else {
      pos = skipField(msg, pos, wt);
    }
  }
  return fields;
}

/** Parse ReadResponse bytes into list of tuple field arrays. */
export function parseReadResponseTuples(data: Uint8Array): unknown[][] {
  const tuples: unknown[][] = [];
  let pos = 0;
  while (pos < data.length) {
    const { value: tag, n: tn } = readVarint(data, pos);
    pos += tn;
    const fn = Number(tag >> 3n);
    const wt = Number(tag & 7n);
    if (fn === 1 && wt === 2) {
      const { slice, nextPos } = readLengthDelimited(data, pos);
      pos = nextPos;
      tuples.push(parseTupleMsg(slice));
    } else {
      pos = skipField(data, pos, wt);
    }
  }
  return tuples;
}

export function decodeReadResponseFirstTuple(raw: Uint8Array): unknown[] | null {
  if (raw.length === 0) return null;
  try {
    const tuples = parseReadResponseTuples(raw);
    if (tuples.length === 0) return null;
    return tuples[0] ?? null;
  } catch {
    return null;
  }
}

export function decodeReadResponseAllTuples(raw: Uint8Array): unknown[][] {
  if (raw.length === 0) return [];
  try {
    return parseReadResponseTuples(raw);
  } catch {
    return [];
  }
}

File: sdks/go/plexspaces/tuplespace_proto_wire.go
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (C) 2025 PlexSpaces Contributors
//
// Protobuf wire encoding/decoding for plexspaces.tuplespace.v1 WriteRequest, ReadRequest,
// and ReadResponse (subset used by WASM host imports). Matches prost output used in
// crates/wasm-runtime simple_component_host and examples/rust/apps/abstractions.

package plexspaces

import (
	"fmt"
	"math"
	"strings"
	"unicode/utf8"
)

func appendVarint(buf []byte, x uint64) []byte {
	for x >= 0x80 {
		buf = append(buf, byte(x)|0x80)
		x >>= 7
	}
	return append(buf, byte(x))
}

func appendFixed64LE(buf []byte, v uint64) []byte {
	for i := 0; i < 8; i++ {
		buf = append(buf, byte(v))
		v >>= 8
	}
	return buf
}

func appendLengthDelimited(buf []byte, fieldNum int, inner []byte) []byte {
	tag := uint64(fieldNum<<3 | 2)
	buf = appendVarint(buf, tag)
	buf = appendVarint(buf, uint64(len(inner)))
	return append(buf, inner...)
}

func encodeTupleField(v any, allowWildcardStar bool) ([]byte, error) {
	var inner []byte
	switch x := v.(type) {
	case nil:
		inner = appendVarint(append([]byte{}, 0x38), 1) // wildcard, field 7
	case string:
		if allowWildcardStar && x == "*" {
			inner = appendVarint(append([]byte{}, 0x38), 1)
			return inner, nil
		}
		if !utf8.ValidString(x) {
			return nil, fmt.Errorf("tuple field string must be valid UTF-8")
		}
		inner = append(inner, 0x1a) // field 3 string
		inner = appendVarint(inner, uint64(len(x)))
		inner = append(inner, x...)
	case bool:
		inner = append(inner, 0x20) // field 4 boolean
		if x {
			inner = appendVarint(inner, 1)
		} else {
			inner = appendVarint(inner, 0)
		}
	case float64:
		if x == math.Trunc(x) && x >= float64(math.MinInt64) && x <= float64(math.MaxInt64) {
			inner = appendVarint(append([]byte{}, 0x08), uint64(int64(x))) // field 1 int64
		} else {
			inner = append(inner, 0x11) // field 2 double
			inner = appendFixed64LE(inner, math.Float64bits(x))
		}
	default:
		return nil, fmt.Errorf("unsupported tuple field type %T", v)
	}
	return inner, nil
}

func encodeTupleFields(tuple []any, allowWildcardStar bool) ([]byte, error) {
	var out []byte
	for _, el := range tuple {
		tf, err := encodeTupleField(el, allowWildcardStar)
		if err != nil {
			return nil, err
		}
		out = appendLengthDelimited(out, 2, tf)
	}
	return out, nil
}

func encodeWriteRequest(tuple []any) ([]byte, error) {
	tupleBody, err := encodeTupleFields(tuple, false)
	if err != nil {
		return nil, err
	}
	return appendLengthDelimited(nil, 2, tupleBody), nil
}

func encodeReadRequest(pattern []any, take bool, maxResults int32) ([]byte, error) {
	// templateBody is a complete Tuple message (repeated TupleField fields = 2 only).
	templateBody, err := encodeTupleFields(pattern, true)
	if err != nil {
		return nil, err
	}
	var out []byte
	out = appendLengthDelimited(out, 2, templateBody)
	if take {
		out = append(out, 0x28) // field 5 take = true
		out = appendVarint(out, 1)
	}
	out = append(out, 0x30) // field 6 max_results
	out = appendVarint(out, uint64(uint32(maxResults)))
	return out, nil
}

func readVarint(data []byte, pos int) (uint64, int, error) {
	var x uint64
	var s uint
	orig := pos
	for i := 0; i < 10; i++ {
		if pos >= len(data) {
			return 0, 0, fmt.Errorf("varint buffer underflow")
		}
		b := data[pos]
		pos++
		if b < 0x80 {
			return x | uint64(b)<<s, pos - orig, nil
		}
		x |= uint64(b&0x7f) << s
		s += 7
	}
	return 0, 0, fmt.Errorf("varint too long")
}

func skipField(data []byte, pos int, wireType uint64) (int, error) {
	switch wireType {
	case 0:
		_, n, err := readVarint(data, pos)
		return pos + n, err
	case 1:
		if pos+8 > len(data) {
			return 0, fmt.Errorf("fixed64 underflow")
		}
		return pos + 8, nil
	case 2:
		ln, n, err := readVarint(data, pos)
		if err != nil {
			return 0, err
		}
		end := pos + n + int(ln)
		if end > len(data) {
			return 0, fmt.Errorf("length-delimited underflow")
		}
		return end, nil
	case 5:
		if pos+4 > len(data) {
			return 0, fmt.Errorf("fixed32 underflow")
		}
		return pos + 4, nil
	default:
		return 0, fmt.Errorf("unknown wire type %d", wireType)
	}
}

func parseTupleFieldMsg(msg []byte) (any, error) {
	pos := 0
	var last any
	for pos < len(msg) {
		tag, n, err := readVarint(msg, pos)
		if err != nil {
			return nil, err
		}
		pos += n
		fn := tag >> 3
		wt := tag & 7
		switch wt {
		case 0:
			v, m, err := readVarint(msg, pos)
			if err != nil {
				return nil, err
			}
			pos += m
			switch fn {
			case 1:
				last = int64(v)
			case 4:
				last = v != 0
			case 6, 7:
				last = nil
			default:
			}
		case 1:
			if pos+8 > len(msg) {
				return nil, fmt.Errorf("double underflow")
			}
			u := uint64(msg[pos]) | uint64(msg[pos+1])<<8 | uint64(msg[pos+2])<<16 | uint64(msg[pos+3])<<24 |
				uint64(msg[pos+4])<<32 | uint64(msg[pos+5])<<40 | uint64(msg[pos+6])<<48 | uint64(msg[pos+7])<<56
			pos += 8
			if fn == 2 {
				last = math.Float64frombits(u)
			}
		case 2:
			ln, m, err := readVarint(msg, pos)
			if err != nil {
				return nil, err
			}
			pos += m
			end := pos + int(ln)
			if end > len(msg) {
				return nil, fmt.Errorf("bytes underflow")
			}
			chunk := msg[pos:end]
			pos = end
			if fn == 3 || fn == 5 {
				last = string(chunk)
			}
		default:
			pos, err = skipField(msg, pos, wt)
			if err != nil {
				return nil, err
			}
		}
	}
	if last == nil {
		return nil, nil
	}
	return last, nil
}

func parseTupleMsg(msg []byte) ([]any, error) {
	pos := 0
	var fields []any
	for pos < len(msg) {
		tag, n, err := readVarint(msg, pos)
		if err != nil {
			return nil, err
		}
		pos += n
		fn := tag >> 3
		wt := tag & 7
		if fn == 2 && wt == 2 {
			ln, m, err := readVarint(msg, pos)
			if err != nil {
				return nil, err
			}
			pos += m
			end := pos + int(ln)
			if end > len(msg) {
				return nil, fmt.Errorf("tuple field underflow")
			}
			sub := msg[pos:end]
			pos = end
			fv, err := parseTupleFieldMsg(sub)
			if err != nil {
				return nil, err
			}
			fields = append(fields, fv)
		} else {
			pos, err = skipField(msg, pos, wt)
			if err != nil {
				return nil, err
			}
		}
	}
	return fields, nil
}

func parseReadResponseTuples(data []byte) ([][]any, error) {
	pos := 0
	var tuples [][]any
	for pos < len(data) {
		tag, n, err := readVarint(data, pos)
		if err != nil {
			return nil, err
		}
		pos += n
		fn := tag >> 3
		wt := tag & 7
		if fn == 2 && wt == 2 {
			ln, m, err := readVarint(data, pos)
			if err != nil {
				return nil, err
			}
			pos += m
			end := pos + int(ln)
			if end > len(data) {
				return nil, fmt.Errorf("tuple message underflow")
			}
			tup, err := parseTupleMsg(data[pos:end])
			if err != nil {
				return nil, err
			}
			pos = end
			tuples = append(tuples, tup)
		} else {
			pos, err = skipField(data, pos, wt)
			if err != nil {
				return nil, err
			}
		}
	}
	return tuples, nil
}

func decodeReadResponseFirstTuple(raw string) ([]any, bool) {
	if raw == "" || strings.HasPrefix(raw, errorPrefix) {
		return nil, false
	}
	tuples, err := parseReadResponseTuples([]byte(raw))
	if err != nil || len(tuples) == 0 {
		return nil, false
	}
	return tuples[0], true
}

func decodeReadResponseAllTuples(raw string) [][]any {
	if strings.HasPrefix(raw, errorPrefix) {
		return nil
	}
	if raw == "" {
		return [][]any{}
	}
	tuples, err := parseReadResponseTuples([]byte(raw))
	if err != nil {
		return nil
	}
	if len(tuples) == 0 {
		return [][]any{}
	}
	return tuples
}
